pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -183,39 +182,92 @@ func (gc *GarbageCollector) Sync(discoveryClient discovery.ServerResourcesInterf
 			return
 		}
 
-		// Something has changed, time to sync.
-		glog.V(2).Infof("syncing garbage collector with updated resources from discovery: %v", newResources)
-
 		// Ensure workers are paused to avoid processing events before informers
 		// have resynced.
 		gc.workerLock.Lock()
 		defer gc.workerLock.Unlock()
 
-		// Resetting the REST mapper will also invalidate the underlying discovery
-		// client. This is a leaky abstraction and assumes behavior about the REST
-		// mapper, but we'll deal with it for now.
-		gc.restMapper.Reset()
-
-		// Perform the monitor resync and wait for controllers to report cache sync.
-		//
-		// NOTE: It's possible that newResources will diverge from the resources
-		// discovered by restMapper during the call to Reset, since they are
-		// distinct discovery clients invalidated at different times. For example,
-		// newResources may contain resources not returned in the restMapper's
-		// discovery call if the resources appeared in-between the calls. In that
-		// case, the restMapper will fail to map some of newResources until the next
-		// sync period.
-		if err := gc.resyncMonitors(newResources); err != nil {
-			utilruntime.HandleError(fmt.Errorf("failed to sync resource monitors: %v", err))
-			return
-		}
-		// TODO: WaitForCacheSync can block forever during normal operation. Could
-		// pass a timeout channel, but we have to consider the implications of
-		// un-pausing the GC with a partially synced graph builder.
-		if !controller.WaitForCacheSync("garbage collector", stopCh, gc.dependencyGraphBuilder.IsSynced) {
-			utilruntime.HandleError(fmt.Errorf("timed out waiting for dependency graph builder sync during GC sync"))
-			return
-		}
+		// Once we get here, we should not unpause workers until we've successfully synced
+		attempt := 0
+		wait.PollImmediateInfinite(time.Second, func() (bool, error) {
```

Comment:
nit: is there any reason to use PollImmediateInfinite instead of PollUntil? I was worried if we would keep looping at some point and never get to check if stop channel was closed. It looks like the stop channel is properly caught in the out `wait.Until`

lgtm otherwise

Reply:
> is there any reason to use PollImmediateInfinite instead of PollUntil?

I didn't want a delay before the first attempt:

> PollUntil always waits interval before the first run of 'condition'.

Reply:
Why one second?

Reply:
In failure cases (discovery, resync, or timeout), we don't want to hotloop

Reply:
I think @roycaihw is right about it looping forever if the stopCh is closed and it continues to encounter discover errors indefinitely. The function inside PollImmediateInfinite will keep returning (false, nil) and the inner loop will never exit. Maybe moving:
```
select {
			case <-stopCh:
				// we're stopping, we don't care that it failed
				return true, nil
			default:
}
```
to be the first thing run within PollImmediateInfinite. Would be better if we had "PollImmediateUntil" but we don't.

Reply:
> I think @roycaihw is right about it looping forever if the stopCh is closed and it continues to encounter discover errors indefinitely. The function inside PollImmediateInfinite will keep returning (false, nil) and the inner loop will never exit

you're right, good catch

> Would be better if we had "PollImmediateUntil" but we don't.

it was easy to add, and mirrors the other *Immediate helpers, so I added that in a separate commit. that let all checking of stopCh fall out of the inner loop.